File: Medium/binary_tree_level_order_traversal/binary_tree_level_order_traversal.py
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def levelOrder(root: Optional[TreeNode]) -> List[List[int]]:
    if not root:
        return []

    result = []
    queue = [root]

    while queue:
        level_size = len(queue)
        current_level = []

        for i in range(level_size):
            node = queue.pop(0)
            current_level.append(node.val)

            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

        result.append(current_level)

    return result

# Example usage:
if __name__ == "__main__":
    print("Testing levelOrder function:")
    # Create a binary tree:
    #         3
    #        / \
    #       9  20
    #          / \
    #         15  7
    root = TreeNode(3)
    root.left = TreeNode(9)
    root.right = TreeNode(20)
    root.right.left = TreeNode(15)
    root.right.right = TreeNode(7)
    print("Input binary tree: ", "[3, 9, 20, Null, Null, 15, 7]")
    print("Level order traversal of the binary tree is:", levelOrder(root))  # Output: [[3], [9, 20], [15, 7]]
    print("")

    # Create another binary tree:
    #         1
    root = TreeNode(1)
    print("Input binary tree: ", "[1]")
    print("Level order traversal of the binary tree is:", levelOrder(root))  # Output: [[1]]
    print("")

    # Create another binary tree:
    #   []
    root = None
    print("Input binary tree: ", "[]")
    print("Level order traversal of the binary tree is:", levelOrder(root))  # Output: []
    print("")

File: Medium/binary_tree_level_order_traversal/binary_tree_level_order_traversal.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
    public:
        vector<vector<int>> levelOrder(TreeNode* root) {
            vector<vector<int>> result;
            if (!root) {
                return result;
            }

            vector<TreeNode*> currentLevel = {root};

            while (!currentLevel.empty()) {
                vector<int> currentValues;
                vector<TreeNode*> nextLevel;

                for (TreeNode* node : currentLevel) {
                    currentValues.push_back(node->val);
                    if (node->left) {
                        nextLevel.push_back(node->left);
                    }
                    if (node->right) {
                        nextLevel.push_back(node->right);
                    }
                }

                result.push_back(currentValues);
                currentLevel = nextLevel;
            }

            return result;
        }
};


// Example usage:
int main() {
    Solution solution1;
    TreeNode* root1 = new TreeNode(3);
    root1->left = new TreeNode(9);
    root1->right = new TreeNode(20);
    root1->right->left = new TreeNode(15);
    root1->right->right = new TreeNode(7);

    vector<vector<int>> result1 = solution1.levelOrder(root1);
    cout << "Testing Binary Tree Level Order Traversal: " << endl;
    cout << "Input: [3, 9, 20, null, null, 15, 7]" << endl;
    cout << "Output: [";
    for (int i = 0; i < result1.size(); i++) {
        cout << "[";
        for (int j = 0; j < result1[i].size(); j++) {
            cout << result1[i][j];  
            if (j != result1[i].size() - 1) {
                    cout << ", ";
                }
        }
        cout << "]";
        if (i != result1.size() - 1) {
            cout << ", ";
        }
    }
    cout << "]" << endl;

    Solution solution2;
    TreeNode* root2 = new TreeNode(1);
    cout << "" << endl;
    vector<vector<int>> result2 = solution2.levelOrder(root2);
    cout << "Input: [1]" << endl;
    cout << "Output: [";
    for (int i = 0; i < result2.size(); i++) {
        cout << "[";
        for (int j = 0; j < result2[i].size(); j++)
        {
            cout << result2[i][j];  
            if (j != result2[i].size() - 1) {
                    cout << ", ";
                }
        }
        cout << "]";
        if (i != result2.size() - 1) {
            cout << ", ";
        }
    }
    cout << "]" << endl;

    Solution solution3;
    TreeNode* root3 = nullptr;
    cout << "" << endl;
    vector<vector<int>> result3 = solution3.levelOrder(root3);
    cout << "Input: []" << endl;
    cout << "Output: [";
    for (int i = 0; i < result3.size(); i++) {
        cout << "[";
        for (int j = 0; j < result3[i].size(); j++)
        {
            cout << result3[i][j];
            if (j != result3[i].size() - 1) {
                    cout << ", ";
                }
        }
        cout << "]";
        if (i != result3.size() - 1) {
            cout << ", ";
        }
    }
    cout << "]" << endl;
    
    return 0;
}